src/iface-opc-ua.c

```diff
@@ -0,0 +1,347 @@
+		read = fread(fileContents.data, sizeof(UA_Byte),
+			     fileContents.length, fp);
+	if (read != fileContents.length)
+		UA_ByteString_clear(&fileContents);
+	} else {
+		fileContents.length = STRING_EMPTY;
+	}
+	fclose(fp);
+
+	return fileContents;
+}
+
+static UA_StatusCode get_node_id(UA_NodeId *parent, int namespace_index,
+			      char *identifier_type,
+			      char *identifier)
+{
+	UA_StatusCode rc = UA_STATUSCODE_GOOD;
+
+	UA_NodeId_init(parent);
+
+	if (!strcmp(identifier_type, NODEIDTYPE_STRING)) {
+		*parent = UA_NODEID_STRING(namespace_index, identifier);
+	} else if (!strcmp(identifier_type, NODEIDTYPE_NUMERIC)) {
+		uint32_t num;
+
+		num = (uint32_t)identifier[0] << 24 |
+		      (uint32_t)identifier[1] << 16 |
+		      (uint32_t)identifier[2] << 8  |
+		      (uint32_t)identifier[3];
+		*parent = UA_NODEID_NUMERIC(namespace_index, num);
+	} else if (!strcmp(identifier_type, NODEIDTYPE_GUID)) {
+		UA_Guid *guid_aux = UA_Guid_new();
+		UA_String string_aux = UA_STRING(identifier);
+
+		UA_Guid_parse(guid_aux, string_aux);
+		*parent = UA_NODEID_GUID(namespace_index, *guid_aux);
+	} else if (!strcmp(identifier_type, NODEIDTYPE_BYTESTRING)) {
+		*parent = UA_NODEID_BYTESTRING(namespace_index, identifier);
+	} else {
+		UA_NodeId_isNull(parent);
+		rc = UA_STATUSCODE_BAD;
+	}
+
+	return rc;
+}
+
+static UA_StatusCode setup_read_data(UA_Variant *value,
+				     struct knot_data_item *data_item)
+{
+	UA_StatusCode rc = UA_STATUSCODE_GOOD;
+	UA_NodeId *parent = UA_NodeId_new();
+
+	UA_Variant_init(value);
+	UA_NodeId_init(parent);
+
+	rc = get_node_id(parent, data_item->namespace,
+			 data_item->identifier_type,
+			 data_item->identifier);
+	if (rc == UA_STATUSCODE_GOOD)
+		rc = UA_Client_readValueAttribute(opc_ua_client,
+						  *parent, value);
+
+	free(parent);
+	return rc;
+}
+
+static UA_StatusCode get_read_value(UA_Variant *value,
+				    uint8_t value_type,
+				    int value_type_size,
+				    union opc_ua_types *tmp)
+{
+	UA_StatusCode rc = UA_STATUSCODE_GOOD;
+
+	switch (value_type) {
+	case KNOT_VALUE_TYPE_BOOL:
+		if (value_type_size == 1)
+			tmp->val_bool = *(UA_Boolean *)value->data;
+		else if (value_type_size == 8)
+			tmp->val_bool = *(UA_SByte *)value->data;
+		else
+			rc = UA_STATUSCODE_BAD;
+		break;
+	case KNOT_VALUE_TYPE_FLOAT:
+		tmp->val_float = *(UA_Float *)value->data;
+		break;
+	case KNOT_VALUE_TYPE_INT:
+		if (value_type_size == 8)
+			tmp->val_i8 = *(UA_Byte *)value->data;
+		else if (value_type_size == 16)
+			tmp->val_i16 = *(UA_Int16 *)value->data;
+		else if (value_type_size == 32)
+			tmp->val_i32 = *(UA_Int32 *)value->data;
+		else
+			rc = UA_STATUSCODE_BAD;
+		break;
+	case KNOT_VALUE_TYPE_UINT:
+		if (value_type_size == 8)
+			tmp->val_u8 = *(UA_SByte *)value->data;
+		if (value_type_size == 16)
+			tmp->val_u16 = *(UA_UInt16 *)value->data;
+		else if (value_type_size == 32)
+			tmp->val_u32 = *(UA_UInt32 *)value->data;
+		else
+			rc = UA_STATUSCODE_BAD;
+		break;
```

Comment:
The need for these conditionals is clear. However, there are too many nested conditionals. Would it be worth extracting the snippets from the KNOT_VALUE_TYPE_BOOL, KNOT_VALUE_TYPE_INT and KNOT_VALUE_TYPE_UIN cases to functions?